src/NHibernate.Test/NHSpecificTest/NH2319/MapFixture.cs

```diff
@@ -0,0 +1,286 @@
+
+			using (var spy = new LogSpy(typeof(QueryPlanCache)))
+			{
+				internalPlanCache.Clear();
+				ShouldBeAbleToFindChildrenByName();
+				AssertFilterPlanCacheMiss(spy);
+			}
+		}
+
+		private const string _filterPlanCacheMissLog = "unable to locate collection-filter query plan in cache";
+
+		private static void AssertFilterPlanCacheHit(LogSpy spy) =>
+			// Each query currently ask the cache two times, so asserting reuse requires to check cache has not been missed
+			// rather than only asserting it has been hit.
+			Assert.That(spy.GetWholeLog(),
+				Contains.Substring("located collection-filter query plan in cache (")
+				.And.Not.Contains(_filterPlanCacheMissLog));
+
+		private static void AssertFilterPlanCacheMiss(LogSpy spy) =>
+			Assert.That(spy.GetWholeLog(), Contains.Substring(_filterPlanCacheMissLog));
+
+		protected override void Configure(Configuration configuration)
+		{
+			configuration.SetProperty("show_sql", "true");
+			configuration.SetProperty("generate_statistics", "true");
+		}
+
+		protected override void OnSetUp()
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				var parent1 = new Parent { Name = "Bob" };
+				_parent1Id = (Guid) session.Save(parent1);
+
+				var parent2 = new Parent { Name = "Martin" };
+				_parent2Id = (Guid) session.Save(parent2);
+
+				var child1 = new Child
+				{
+					Name = "Jack",
+					Parent = parent1
+				};
+				_child1Id = (Guid) session.Save(child1);
+				parent1.ChildrenMap.Add(child1.Id, child1);
+
+				var child2 = new Child
+				{
+					Name = "Piter",
+					Parent = parent1
+				};
+				session.Save(child2);
+				parent1.ChildrenMap.Add(child2.Id, child2);
+
+				var grandChild1 = new GrandChild
+				{
+					Name = "Kate",
+					Child = child2
+				};
+				session.Save(grandChild1);
+				child2.GrandChildren.Add(grandChild1);
+
+				var grandChild2 = new GrandChild
+				{
+					Name = "Mary",
+					Child = child2
+				};
+				session.Save(grandChild2);
+				child2.GrandChildren.Add(grandChild2);
+
+				var child3 = new Child
+				{
+					Name = "Jack",
+					Parent = parent2
+				};
+				_child3Id = (Guid) session.Save(child3);
+				parent2.ChildrenMap.Add(child1.Id, child3);
+
+				session.Flush();
+				transaction.Commit();
+			}
+		}
+
+		protected override void OnTearDown()
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				session.Delete("from System.Object");
+
+				session.Flush();
+				transaction.Commit();
+			}
+		}
+
+		protected override HbmMapping GetMappings()
+		{
+			var mapper = new ModelMapper();
+			mapper.Class<Parent>(rc =>
+			{
+				rc.Id(x => x.Id, m => m.Generator(Generators.GuidComb));
+				rc.Property(x => x.Name);
+				rc.Map(x => x.ChildrenMap,
+					map => map.Cascade(Mapping.ByCode.Cascade.All | Mapping.ByCode.Cascade.DeleteOrphans),
```

Comment:
I have not found how to specify the index, although we can specify it in hbm. Missing feature of by code mapping? Anyway, works with defaults for the test.

Reply:
It is supposed that you use the `Map` overload which takes 4 lambdas, but `IMapKeyRelation` is missing the index part.

Reply:
Or probably ` Element` will behave as you need.